riscv/mmu.h

```diff
@@ -342,6 +342,42 @@ class mmu_t
 #endif
   }
 
+  void set_target_big_endian(bool enable)
+  {
+#ifdef RISCV_ENABLE_DUAL_ENDIAN
+    target_big_endian = enable;
+#else
+    assert(enable == false);
+#endif
```

Comment:
```
#ifdef` RISCV_ENABLE_DUAL_ENDIAN
    target_big_endian = enable;
#endif
```

Reply:
Should probably keep the assert in the #else branch here, otherwise you end up with a function which exists but silently does nothing when called.

Reply:
What I expect is 
`diff --git a/riscv/mmu.cc b/riscv/mmu.cc
index db90342..ad48f4c 100644
--- a/riscv/mmu.cc
+++ b/riscv/mmu.cc
@@ -6,9 +6,7 @@
 
 mmu_t::mmu_t(simif_t* sim, processor_t* proc)
  : sim(sim), proc(proc),
-#ifdef RISCV_ENABLE_DUAL_ENDIAN
   target_big_endian(false),
-#endif
   check_triggers_fetch(false),
   check_triggers_load(false),
   check_triggers_store(false),
diff --git a/riscv/mmu.h b/riscv/mmu.h
index e42ac82..2670d9b 100644
--- a/riscv/mmu.h
+++ b/riscv/mmu.h
@@ -346,36 +346,22 @@ public:
   {
 #ifdef RISCV_ENABLE_DUAL_ENDIAN
     target_big_endian = enable;
-#else
-    assert(enable == false);
 #endif
   }
 
   bool is_target_big_endian()
   {
-#ifdef RISCV_ENABLE_DUAL_ENDIAN
     return target_big_endian;
-#else
-    return false;
-#endif
   }
 
   template<typename T> inline T from_target(target_endian<T> n) const
   {
-#ifdef RISCV_ENABLE_DUAL_ENDIAN
     return target_big_endian? n.from_be() : n.from_le();
-#else
-    return n.from_le();
-#endif
   }
 
   template<typename T> inline target_endian<T> to_target(T n) const
   {
-#ifdef RISCV_ENABLE_DUAL_ENDIAN
     return target_big_endian? target_endian<T>::to_be(n) : target_endian<T>::to_le(n);
-#else
-    return target_endian<T>::to_le(n);
-#endif
   }
 
 private:
@@ -462,6 +448,8 @@ private:
 
 #ifdef RISCV_ENABLE_DUAL_ENDIAN
   bool target_big_endian;
+#else
+  const bool target_big_endian;
 #endif
   bool check_triggers_fetch;
   bool check_triggers_load;
`

Reply:
If you do it like that then the compiler has no way to optimize away the runtime check.  It only knows that `target_big_endian` is constant, not what the actual constant value _is_ (that must be set by the constructor, which lives in a different compilation unit, and it's also possible for different constructors to set different values).

In fact, even using `const bool target_big_endian = false` is not enough to get the optimization (just checked with clang), it needs to be `static const bool target big_endian = false`.

Of course, if you are fine with not optimizing and always having the check done at runtime regardless of `--enable-dual-endian` or not, then I have no problem with that either.  In that case the `#ifdef` around the declation can be dropped and it can always be `bool target_big_endian`.

Reply:
We definitely want to optimize these checks away for `--disable-dual-endian`.